### ISSUE TITLE
Cannot reuse same history list with different slices on different indicators

### ISSUE THREAD

**Issue by MithrilMan** [DaveSkender/Stock.Indicators#170]
Hello
I'm trying your library (thanks for the effort so far) and I've found a bug regarding the usage of the quotes to generate different indicators using the same set of historical quotes but taking a different slice of them

Debug was hard on my side, so I took a look at your code and I see that your model classes have an internal Index that you use and I think that's the problem.

Suppose I get historical data from my POCO classes and convert them to your `Quote` model
suppose I've a list `List<Quote> history` of 100 `Quote`s

Now if I call
```
Indicator.GetBollingerBands(history.TakeLast(20), 20);
Indicator.GetRsi(history.TakeLast(14), 14);
```

I get a
`System.ArgumentOutOfRangeException: 'Index was out of range. Must be non-negative and less than the size of the collection.'`

I think the problem is that if I reuse Quote objects taking a different slice of the array, the bugs pops out because there is the Index field inconsistence inside your model classes

This means that historical quotes cannot be reused this way but I must always use the same set and I think it's not good because of performance (I'm not interested in applying indicators on the whole historical set in my scenario

Thanks

**Comment by MithrilMan** [comment]
here a screenshot to show you current data that gets passed to the second indicator
as you see, first item starts from Index 7 (and I don't have a way to reset these indexes I think)

![image](https://user-images.githubusercontent.com/5107375/98615203-59e0b880-22fa-11eb-97b2-493fcd111501.png)

**Comment by DaveSkender** [comment]
Yah, the Index number itself is important for locating the start of the dataset relative to the lookback period.  Using TakeLast cuts it off unnaturally.  Since history was designed for reusability, it doesn’t recompose the Index (for performance reasons).

For now, you might try to cut the results down instead of the input history.

Let me think about it.  There’s likely both another workaround, so you can cut it like you depict, and perhaps a bug fix where we detect bad indexes and do a recompose.  I’ll also consider options for dropping the Index altogether as well — can probably find a better way.

**Comment by MithrilMan** [comment]
> Let me think about it. There’s likely both another workaround, so you can cut it like you depict, and perhaps a bug fix where we detect bad indexes and do a recompose. I’ll also consider options for dropping the Index altogether as well — can probably find a better way.

An easy way before implementing a proper solution could be to implement an extension of `IEnumerable<Quote>` in your library that reset the indexes, so external library can call it

that way I can use that extension like `history.Reset()` or something like that

**Comment by DaveSkender** [comment]
Another concern is that if you cut down history, you might not be able to restore it to a larger slice.

It may take me a few days to review, but will probably get to it this weekend.  If you want to submit a PR with a recommended fix, that’s also appreciated, only if your up to it.

**Comment by MithrilMan** [comment]
I see if I'll find time tomorrow after my daily job (3 am here), I need to take time to find the right place to put that extension on your codebase

Note that I'm not cutting my history.
My original history list still contains full 100 items, I just want to compute indicators on a subset and so I use history.TakeLast
The use case makes sense, anyway take your time for your considerations about it.
Thanks.

**Comment by DaveSkender** [comment]
If it’s not permanently cutting, I’ll probably just check the first Index number in the dataset in the Cleaner method.  If it’s not 1, redo the Index.

**Comment by DaveSkender** [comment]
I think this fix will work fine.  I need to add some unit tests to confirm when I have more time, and not writing code from my iPad :)

**Comment by DaveSkender** [comment]
As a side note, I’d recommend using more history for RSI anyway.  It uses EMA under the hood, which has a convergence algorithm.  Less history means less precision.  If you compare 14 periods vs 250 periods of history for RSI(14) you’ll see small decimal point differences.  Not much of an issue for most use cases, but worth pointing out.

**Comment by MithrilMan** [comment]
thanks for your tip, I'm now using 250 values to try it out
However I'm facing a problem that I don't know if it's by design or not:
In my scenario, the last item of the history is the current candle that of course change in real time
Even changing the date and value of the last historical Quote, the result of the last RSI value doesn't change, here a proof

![image](https://user-images.githubusercontent.com/5107375/98696463-1f1b6680-2374-11eb-8ce7-8175befa9f1b.png)

Is it normal?
Financial chart has indicators that change their current last point in real time, that's the same I'd need to implement

Thanks

**Comment by DaveSkender** [comment]
I've added a fix to recompose the index if it detects a bad starting value.

I'm also adding an undocumented history.RemoveIndex() extension that will allow a user to remove the internal index from a previously utilized or cleaned history. With this fix, you would not need this extension; however, I'm adding it in case there are unidentified reasons for a user to need to remove the internal indexing.

This fix is available in version 1.0.6

**Comment by MithrilMan** [comment]
cool thanks will test it

**Comment by MithrilMan** [comment]
ok I tested latest version (1.1.2) and the fix doesn't work

![image](https://user-images.githubusercontent.com/5107375/99193648-aa349c00-277a-11eb-9a3b-387732efdf30.png)

In your attempt to fix it, you check the wrong condition because as you see, when I use just a subset of my historical data and then try to use again the full history, I'm in a situation like the screenshot where I have anotehr Index = 1 in between, that of course make it crash.

It's not enough to check if first index is 1 because you have to ensure that there aren't other index = 1 in between
probably is enough to check that the last element has the expected index to not have to iterate each item
basically add the condition that last item index should be equal to the historyList count

You can add a simple test as computing an indicator 3 times: full set, reduced set, full set

```
IEnumerable<RocResult> roc = Indicator.GetRoc(Data, 14);
roc = Indicator.GetRoc(Data.TakeLast(15), 14);
roc = Indicator.GetRoc(Data, 14);
```

**Comment by MithrilMan** [comment]
also, what's the need of Index?
I think it's overkilling and if there is really a need for a class to have an Index within (but please explain me why) then would be better to have a IQuote with High,Low,Close etc... and then an IndexedQuote that has an additional Index property only for indicators that really need it

but if we take for example the RSI implementation that uses the Index property, to me it can just make use of the for iteration variable `i`

https://github.com/DaveSkender/Stock.Indicators/blob/873fc24412f5c73daf658cf8238f4cf475e49dd8/indicators/Rsi/Rsi.cs#L38

**Comment by DaveSkender** [comment]
Okay, I'll check that scenario.  I had a unit test to check for corruption and thought I had fixed it.  As a side note, I already have a plan (AB#763) to review and probably remove the use of Index internally.  I just need to see if I can reasonably do that without causing other problems.

**Comment by MithrilMan** [comment]
> Okay, I'll check that scenario. I had a unit test to check for corruption and thought I had fixed it. As a side note, I already have a plan ([AB#763](https://dev.azure.com/skender/5123ca47-74f2-4d67-a5d4-c4d90b8d670a/_workitems/edit/763)) to review and probably remove the use of Index internally. I just need to see if I can reasonably do that without causing other problems.
> 
> For now, I'd avoid trying to mess with `history` after it's composed and used or just use that helper `RemoveIndex`. There's really not much of a performance improvement to do this micro-resizing between uses. It's not how I had intended in the design. I get that you're trying to hyper-optimize, but simply using `TakeLast` is probably a higher performance hit than the benefit from doing it.

I'm actually appending `RemoveIndex()` wherever I pass Data, Takelast is not a performance hit at all because prevent to compute lot of data in indicators
I'm actually monitoring real time 50+ symbols and my history buffer is 288 elements (1 day with 5 min candles)
there are indicators that I can use by just feeding them with a subset of items, e.g. if I'm interested only in current RoC value, I can feed Roc indicator with 14+1 samples on a loopback of 14, saving 288-15 computation of that indices (that I've to do multiple times per second on all symbols)

**Comment by DaveSkender** [comment]
I think for this one, it's probably best for me to just revert the fix.  It's better to get an error in this use case than to get corrupted data without an error.  I'll leave the `RemoveIndex` helper so you'll at least be un-stuck.

**Comment by MithrilMan** [comment]
fine

**Comment by DaveSkender** [comment]
I'm closing this issue since it is remediated with the `RemoveIndex` workaround and with the understanding that it will be resolved more comprehensively in #180
